File: lib/dev_spectrogram/xspectrogram.m
function qmesh = xspectrogram(x, y, Ts, Nest, resolution, vmin, vmax, ax)

    % check for input arguments and set defaults if not provided
    if nargin < 4 || isempty(Nest)
        Nest = 200;
    end
    if nargin < 5 || isempty(resolution)
        resolution = 200;
    end
    if nargin < 6 || isempty(vmin)
        vmin = 1e-1;
    end
    if nargin < 7 || isempty(vmax)
        vmax = 1e1;
    end
    if nargin < 8 || isempty(ax)
        ax = gca; % use current axes if none provided
    end

    % Todo: check if it is usefull to remove mean here
    % x = x - mean(x);
    % y = y - mean(y);

    Ndata = size(x, 1);

    % calculate overlap size for windowing
    Noverlap = floor(0.9 * Nest);

    % find the minimum and maximum values of y and generate linearly spaced vector for the output y-axis
    ymin = min(y);
    ymax = max(y);
    dy = (ymax - ymin) / resolution;
    y_axis = (ymin:dy:ymax-dy).';

    % compute the frequency bins for the output x-axis
    % freq = (0:Nest/2).' * (1 / (Nest * Ts));
    df = 1/(Nest*Ts);
    freq = (0:df:1/Ts-df).';

    % initialize matrices for the spectrogram data and counter for averaging
    Pavg = zeros(resolution, Nest);
    Navg = zeros(resolution, 1);

    % generate the windowing function
    window = hann(Nest);

    % factor 2 so that the magnitude corresponds to a single sided spectrum
    % 2.3*sin(2*pi*f0*time) <=> sqrt(puu(f0)) = 2.3
    W = sum(window) / Nest / 2;

    ind_start = 1;
    ind_end = Nest;
    Ndelta  = Nest - Noverlap;

    % process the signal in chunks
    while ind_end <= Ndata

        % calculate the start and end indices for the y values
        y_ind_start = round(ind_start + (Nest - Ndelta)/2);
        y_ind_end   = round(ind_start + (Nest + Ndelta)/2);

        % map y values to indices in the spectrogram resolution
        indices = round((y(y_ind_start:y_ind_end) - ymin) / (ymax - ymin) * (resolution - 1)) + 1;

        ind = ind_start:ind_end;

        inp_act = x(ind);

        inp_act = window .* inp_act;

        % Todo: check if it is usefull to remove mean here
        % inp_act = inp_act - mean(inp_act);

        U = fft(inp_act) / (Nest * W);
        Pact = U .* conj(U);

        % accumulate the magnitudes into the spectrogram matrix
        for i = 1:length(indices)
            idx = indices(i);
            Pavg(idx, :) = Pavg(idx, :) + Pact.';
            Navg(idx) = Navg(idx) + 1;
        end

        ind_start = ind_start + Ndelta;
        ind_end   = ind_end   + Ndelta;
    end


    ind_Navg = Navg ~= 0;
    Pavg = Pavg(ind_Navg,:) ./ Navg(ind_Navg);

    ind = freq <= 1 / (2 * Ts);
    freq = freq(ind);
    Pavg = Pavg(:, ind);

    % Plot the spectrogram data as a pseudocolor plot
    qmesh = pcolor(ax, freq, y_axis, sqrt(Pavg));
    set(qmesh, 'EdgeColor', 'none'); % Remove gridlines
    clim(ax, [vmin vmax]); % Set the color axis scaling
    c = colorbar(ax); % Add a colorbar
    set(ax, 'ColorScale', 'log');
    xlabel(ax, 'Frequency (Hz)'); % Label the y-axis
end

File: lib/dev_spectrogram/xspectrogram.py
def xspectrogram(x: list[float], y: list[float], Fs: float, NFFT: int = 200, resolution: int = 201, vmin: float = -50, vmax: float = 0, ax: plt.Axes = None, **kwargs):

    if ax is None:
        ax = plt.gca()

    noverlap: int = int(2/3 * NFFT)
    ymin: float = np.min(y)
    ymax: float = np.max(y)

    n = np.linspace(ymin, ymax, resolution)
    f = np.fft.rfftfreq(NFFT, 1/Fs)
    nf = len(f)

    spec: np.ndarray = np.zeros((resolution, nf), dtype=float)
    specCounter: np.ndarray = np.zeros(resolution, dtype=int)
    window: np.ndarray = np.hanning(NFFT)    

    ptr: int = 0
    ptrEnd: int = NFFT
    ptrDelta: int = int(NFFT - noverlap)

    while ptrEnd < len(x):

        yStart: int = int(ptr + (NFFT - ptrDelta)/2)
        yEnd: int = int(ptr + (NFFT + ptrDelta)/2)
        indices: list[int] = [round((y - ymin) / (ymax - ymin) * (resolution - 1)) for y in y[yStart:yEnd]]

        xWindowed: list[float] = np.multiply(window, x[ptr:ptrEnd])
        xMag: list[float] = np.abs(np.fft.rfft(xWindowed))

        for i in indices:
            spec[i] = np.add(spec[i], xMag)
            specCounter[i] += 1

        ptr += ptrDelta
        ptrEnd += ptrDelta

    for i in range(resolution):
        if specCounter[i] != 0:
            # psd = 20 * log10(2 * magnitude / bandwidth)
            spec[i] = 20 * np.log10(np.divide(spec[i], specCounter[i] * Fs/4))
        else:
            spec[i] = [vmin] * nf

    qmesh = ax.pcolormesh(n, f, np.transpose(spec), vmin=vmin, vmax=vmax, **kwargs)
    plt.colorbar(mappable=qmesh, ax=ax, pad=0.02, label='PSD (dB/Hz)')

    ax.set_ylabel('Frequency (Hz)')

    return qmesh